### PR TITLE
[JS/WebGPU] Avoid producing presentKey/presentValue outputs if pastKey/pastValue …

### PR DESCRIPTION
**Pull request by satyajandhyala**
Avoid producing presentKey/presentValue outputs if pastKey/pastValue don't exists.

### CODE REVIEW

**Comment by tianleiwu** [comment]
For the prompt, there is no past_key or past_value, however present_key and present_value exist.

In theory, we can use empty past_key, where its shape has 0 in the past_sequence_length dimension. Does JS/WebGPU support that?

**Comment by satyajandhyala** [comment]
> For the prompt, there is no past_key or past_value, however present_key and present_value exist.
> 
> In theory, we can use empty past_key, where its shape has 0 in the past_sequence_length dimension. Does JS/WebGPU support that?

No. Empty input tensors are not supported.

**Comment by satyajandhyala** [comment]
> > For the prompt, there is no past_key or past_value, however present_key and present_value exist.
> > In theory, we can use empty past_key, where its shape has 0 in the past_sequence_length dimension. Does JS/WebGPU support that?
> 
> No. Empty input tensors are not supported.

If necessary we need to change logic in MHA.

**Comment by satyajandhyala** [comment]
> > > For the prompt, there is no past_key or past_value, however present_key and present_value exist.
> > > In theory, we can use empty past_key, where its shape has 0 in the past_sequence_length dimension. Does JS/WebGPU support that?
> > 
> > 
> > No. Empty input tensors are not supported.
> 
> If necessary we need to change logic in MHA.

Added code to accept empty pastKey/pastValue and produce presentKey/presentValue.

File: js/web/lib/wasm/jsep/webgpu/ops/multihead-attention.ts [hunk 1]
```diff
@@ -85,7 +85,7 @@ const validateInputs = (inputs: readonly TensorView[], attributes: AttentionAttr
   let pastSequenceLength = 0;
   let maxSequenceLength = 0;
   const headSize = Math.floor(hiddenSize / attributes.numHeads);
-  if (pastKey && pastValue) {
+  if (pastKey && pastValue && ShapeUtil.size(pastKey.dims) && ShapeUtil.size(pastValue.dims)) {
```
Comment (tianleiwu) [inline js/web/lib/wasm/jsep/webgpu/ops/multihead-attention.ts:88]: Might not need such change.

The original check is fine when past_key has shape like [B, N, P, H] when P === 0 (Need some tests to see whether JS supports such shape).

**Comment by tianleiwu** [comment]
The only concern is that it might break backward-compatible.

For Phi-3-mini-4k-instruct-onnx-web, it is fine since the model has both past and present.

@guschmue, @fs-eire do you know whether there are other LLMs onnx models supported by WebGPU might encounter issue?

**Comment by guschmue** [comment]
I can test it for sure with phi3, gemma2, tinyllama and smollm

**Comment by guschmue** [comment]
phi3, gemma2-2b, smollm, stablellm and tinyllama are happy with the change. Last 2 are via optimum, the others are mb.

**Review by guschmue — APPROVED** [review]

**Pull request merged**